client/app/pages/queries/QuerySource.jsx

```diff
@@ -1,19 +1,123 @@
 import PropTypes from 'prop-types';
 import { react2angular } from 'react2angular';
+import Select from 'antd/lib/select';
 import { routesToAngularRoutes } from '@/lib/utils';
 import { Query } from '@/services/query';
+import { DataSource, SCHEMA_NOT_SUPPORTED } from '@/services/data-source';
+import notification from '@/services/notification';
+import recordEvent from '@/services/recordEvent';
 
 import QueryPageHeader from './components/QueryPageHeader';
+import SchemaBrowser from './components/SchemaBrowser';
+
+import './query-source.less';
+
+function getSchema(dataSource, refresh = undefined) {
+  if (!dataSource) {
+    return Promise.resolve([]);
+  }
+
+  return dataSource.getSchema(refresh)
+    .then((data) => {
+      if (data.schema) {
+        return data.schema;
+      } else if (data.error.code === SCHEMA_NOT_SUPPORTED) {
+        return [];
+      }
+      return Promise.reject(new Error('Schema refresh failed.'));
+    })
+    .catch(() => {
+      notification.error('Schema refresh failed.', 'Please try again later.');
+      return Promise.resolve([]);
+    });
+}
+
+function QuerySource(props) {
+  const [query, setQuery] = useState(props.query);
+  const [dataSources, setDataSources] = useState([]);
+  const dataSource = useMemo(() => (find(dataSources, { id: query.data_source_id }) || null), [query, dataSources]);
+  const [schema, setSchema] = useState([]);
+  const refreshSchemaTokenRef = useRef(null);
+
+  useEffect(() => {
+    recordEvent('view_source', 'query', query.id);
+
+    let isCancelled = false;
+    DataSource.query().$promise.then((data) => {
+      if (!isCancelled) {
+        setDataSources(data);
+      }
+    });
+
+    return () => { isCancelled = true; };
+  }, []);
+
+  const reloadSchema = useCallback((refresh = undefined) => {
+    const refreshToken = Math.random().toString(36).substr(2);
+    refreshSchemaTokenRef.current = refreshToken;
+    getSchema(dataSource, refresh).then((data) => {
+      if (refreshSchemaTokenRef.current === refreshToken) {
+        setSchema(data);
+      }
+    });
+  }, [dataSource]);
+
+  useEffect(() => {
+    reloadSchema();
+  }, [reloadSchema]);
+
+  useEffect(() => (() => {
+    // cancel pending operations
+    refreshSchemaTokenRef.current = null;
+  }), []);
+
+  const handleDataSourceChange = useCallback((dataSourceId) => {
+    const newQuery = clone(query);
+    newQuery.data_source_id = dataSourceId;
+    setQuery(newQuery);
+  }, [query]);
 
-function QuerySource({ query }) {
   return (
     <div className="query-page-wrapper">
       <div className="container">
         <QueryPageHeader query={query} sourceMode />
       </div>
       <main className="query-fullscreen">
-        Content
+        <nav>
+          <div className="editor__left__data-source">
+            <Select
+              className="w-100"
+              value={query.data_source_id}
+              disabled={!query.can_edit}
+              onChange={handleDataSourceChange}
```

Comment:
```suggestion
              onChange={handleDataSourceChange}
              optionFilterProp="children"
              showSearch
```
(Hope that alone will allow the search)

Reply:
Done